service/authorization/authorization.go

```diff
@@ -359,40 +360,107 @@ func (as *AuthorizationService) GetDecisions(ctx context.Context, req *authoriza
+	rule := attribute.GetRule()
+	for _, v := range attribute.GetValues() {
+		// if scope exists and current attribute value FQN is not in scope
+		if !(scopeMap == nil || scopeMap[v.GetFqn()]) {
+			// skip
+			continue
+		}
+		// trim attribute values (by default only keep single value relevant to FQN)
+		// This is key to minimizing the rego query size.
+		values := []*policy.Value{v}
+		if rule == policy.AttributeRuleTypeEnum_ATTRIBUTE_RULE_TYPE_ENUM_HIERARCHY {
+			// restore ALL attribute values if attribute rule is hierarchical
+			// This is key to honoring comprehensive hierarchy.
+			values = attribute.GetValues()
 		}
-		request.Fqns = attributeFqns
-	} else {
-		// get subject mappings
-		request.Fqns = req.GetScope().GetAttributeValueFqns()
+		// only clone relevant fields for attribute
+		a := &policy.Attribute{Rule: rule, Values: values}
+		fqnAttrVals[v.GetFqn()] = &attr.GetAttributeValuesByFqnsResponse_AttributeAndValue{Attribute: a, Value: v}
+	}
+	return fqnAttrVals
+}
+
+// makeValsByFqnsLookup creates a lookup map of attribute values by FQN.
+func makeValsByFqnsLookup(attrs []*policy.Attribute, subMapsByVal map[string][]*policy.SubjectMapping, scopeMap map[string]bool) map[string]*attr.GetAttributeValuesByFqnsResponse_AttributeAndValue {
+	fqnAttrVals := make(map[string]*attr.GetAttributeValuesByFqnsResponse_AttributeAndValue)
+	for i := range attrs {
+		// add subject mappings to attribute values
+		attrs[i].Values = updateValsWithSubMaps(attrs[i].GetValues(), subMapsByVal)
+		// update the lookup map with attribute values by FQN
+		fqnAttrVals = updateValsByFqnLookup(attrs[i], scopeMap, fqnAttrVals)
+	}
+	return fqnAttrVals
+}
+
+// makeScopeMap creates a map of attribute value FQNs.
+func makeScopeMap(scope *authorization.ResourceAttribute) map[string]bool {
+	// if scope not defined, return nil pointer
+	if scope == nil {
+		return nil
+	}
+	scopeMap := make(map[string]bool)
+	// add attribute value FQNs from scope to the map
+	for _, fqn := range scope.GetAttributeValueFqns() {
+		scopeMap[fqn] = true
+	}
+	return scopeMap
+}
+
+func (as *AuthorizationService) GetEntitlements(ctx context.Context, req *authorization.GetEntitlementsRequest) (*authorization.GetEntitlementsResponse, error) {
+	as.logger.DebugContext(ctx, "getting entitlements")
+	attrsRes, err := as.sdk.Attributes.ListAttributes(ctx, &attr.ListAttributesRequest{})
+	if err != nil {
+		as.logger.ErrorContext(ctx, "failed to list attributes", slog.String("error", err.Error()))
+		return nil, status.Error(codes.Internal, "failed to list attributes")
 	}
-	avf, err := as.sdk.Attributes.GetAttributeValuesByFqns(ctx, &request)
+	subMapsRes, err := as.sdk.SubjectMapping.ListSubjectMappings(ctx, &subjectmapping.ListSubjectMappingsRequest{})
```

Comment:
@suchak1 based on @strantalis' feedback would this provide any additional performance since it reduces the subject mappings to the scope via the DB request? https://github.com/opentdf/platform/blob/main/service/policy/db/subject_mappings.go#L766

Reply:
See [this answer](https://github.com/opentdf/platform/pull/1271#issuecomment-2276804350).

Reply:
also we'd need to call MatchSubjectMappings from within the rego but rego doesnt doesnt currently support grpc

Reply:
Note `MatchSubjectMappings` needs to be updated regardless: https://github.com/opentdf/platform/issues/1290

Reply:
We now get the entity representation outside of rego. So once match subject mappings is updated we can call it before rego.